nsxt/resource_nsxt_manager_cluster.go

```diff
@@ -0,0 +1,446 @@
+}
+
+func getHostIPFromClusterConfig(client interface{}, clusterConfig nsxModel.ClusterConfig) (string, error) {
+	c := client.(nsxtClients)
+	host := c.Host
+	host = strings.TrimPrefix(host, "https://")
+	// Check if host is ip address or fqdn, if it's ip address then we are all set,
+	// if it's fqdn, we loop through clusterconfig, find the node with same fqdn and
+	// return its ip address
+	ip := net.ParseIP(host)
+	isIPAddress := ip != nil
+	if isIPAddress {
+		return host, nil
+	}
+	nodes := clusterConfig.Nodes
+	for _, node := range nodes {
+		fqdn := *node.Fqdn
+		if fqdn == host {
+			if node.ApiListenAddr != nil {
+				// return ipv4 address if it has one, if it doesn't return ipv6 address
+				addr := node.ApiListenAddr
+				if *addr.IpAddress != "" {
+					return *addr.IpAddress, nil
+				}
+				return *addr.Ipv6Address, nil
+			}
+			v4, v6 := getHTTPSIPFromNodeEntity(node.Entities)
+			if v4 != "" {
+				return v4, nil
+			}
+			return v6, nil
+		}
+	}
+	return "", fmt.Errorf("Failed to get host ip from cluster config")
+}
+
+func getHTTPSIPFromNodeEntity(entities []nsxModel.NodeEntityInfo) (string, string) {
+	// returns the ipv4 and ipv6 address of a node's https port 443
+	for _, entity := range entities {
+		if *entity.Port == 443 {
+			return *entity.IpAddress, *entity.Ipv6Address
+		}
+	}
+	return "", ""
+}
+
+func getNewNsxtClient(node NsxClusterNode, d *schema.ResourceData, clients interface{}) (interface{}, error) {
+	// The API of joining a new node to cluster needs to be called with the new node's ip,
+	// to do that we initialize a new nsxtClient with new node's ip as its host
+	username := node.UserName
+	password := node.Password
+	host := node.IPAddress
+	if !strings.HasPrefix(host, "https://") {
+		host = fmt.Sprintf("https://%s", host)
+	}
+	c := clients.(nsxtClients)
+	commonConfig := c.CommonConfig
+	newClients := nsxtClients{
+		CommonConfig: commonConfig,
+	}
+	err := configureNewClient(&newClients, &c, host, username, password)
+	if err != nil {
+		return nil, err
+	}
+	return newClients, nil
+}
+
+func configureNewClient(newClient *nsxtClients, oldClient *nsxtClients, host string, username string, password string) error {
+	newClient.Host = host
+	securityCtx, err := getConfiguredSecurityContext(newClient, "", "", "", username, password)
+	if err != nil {
+		return fmt.Errorf("Failed to configure new client with host %s: %s", host, err)
+	}
+	newClient.PolicySecurityContext = securityCtx
+	newClient.PolicyHTTPClient = oldClient.PolicyHTTPClient
+	err = initNSXVersion(getPolicyConnector(*newClient))
+	if err != nil {
+		return fmt.Errorf("Failed to configure new client with host %s: %s", host, err)
+	}
+	return nil
+}
+
+func joinNodeToCluster(clusterID string, certSha256Thumbprint string, guestNode NsxClusterNode, masterNodeIP string, d *schema.ResourceData, m interface{}) error {
+	c, err := getNewNsxtClient(guestNode, d, m)
+	if err != nil {
+		return err
+	}
+	newNsxClients := c.(nsxtClients)
```

Comment:
I think log printouts are required for each joining node or node being removed. I would suggest `info` level logs for beginning/end of the process.

Reply:
Thanks. Logs are still missing in remove workflow.

Reply:
Got it